src/coreclr/jit/valuenum.cpp

```diff
@@ -8830,22 +8831,13 @@ void Compiler::fgValueNumberTree(GenTree* tree)
 
                 if (!wasNewobj)
                 {
-
                     // Is this invariant indirect expected to always return a non-null value?
+                    // TODO-VNTypes: non-null indirects should only be used for TYP_REFs.
                     if ((tree->gtFlags & GTF_IND_NONNULL) != 0)
```

Comment:
I guess NONNULL can also be used as a sign that target is never zero so it's not necessary to be TYP_REF?

Reply:
Yes.

The intent of this comment was the following:

1) This code is broken right now for any indirects that may have types that do not have 1:1 relationship with their addresses, because the VN for `IND(short addr)` will have be same VN as for `IND(int addr)`, that kind of thing. That's what the first `TODO` tracks. However, we are safe right now because no cases exist where we can have such indirections here.
> Why?
2) The code which checks for readonly-init-ness of fields in `fgMorphField` will only ever deduce it for `ref`-typed fields, because that's what the VM implementation for `getCurrentClass` checks for.

3) Were that to no longer become true, we could have cases where we will be producing `IND(int addr)` trees marked as `NONNULL` in `fgMorphField`, while their actual values could be zero. Obviously, this would be a bug in `fgMorphField`, but that is what made me think of this scenario.

4) It is cheaper (in terms of TP) to leave the non-null indirects with one arg only (the address) and not have to encode the result type, if we don't have scenarios (right now) which will need this result type, and instead assert their `ref`/byref`-ness, as those types always have 1:1 relationship with their addresses (well, `byref`s are too a little shaky...).

So, all these concerns combined, lead me to write this comment.